### PR TITLE
feat (sections): #3423 implement deduping for sections

### PR DESCRIPTION
**Pull request by rlr**
@csadilek is this what you were thinking? f? Still need to write tests.

### CODE REVIEW

**Comment by csadilek** [comment]
@rlr Nice! Yes, I think that is a much better solution as it will work for any other feed/section brought in later. Feeds can simply declare what feeds to dedupe from. In most cases this will also prevent rendering cards and immediately updating them as this happens in updateSection now.

So, 👍 :)

**Comment by rlr** [comment]
@csadilek ok, got a couple of tests added and passing. It was another test with a side effect that had me pulling my hair out. r?

File: system-addon/lib/SectionsManager.jsm [hunk 1]
```diff
@@ -123,9 +127,30 @@ const SectionsManager = {
...
+    }
+  },
+  dedupeRows(id, options) {
+    const dedupeFrom = this.sections.get(id).dedupeFrom;
+    if (dedupeFrom && dedupeFrom.length > 0 && options.rows) {
+      for (const sectionId of dedupeFrom) {
+        const section = this.sections.get(sectionId);
+        if (section && section.rows) {
+          const [, newRows] = this.dedupe.group(section.rows, options.rows);
+          options.rows = newRows;
```
Comment (piatra) [inline system-addon/lib/SectionsManager.jsm]: Could we avoid mutating the arguments?

**Comment by piatra** [comment]
There is an issue I ran into when rebasing over this patch. I will try to describe the flow:
1. You bookmark a topstory
2. HighlightsFeed listens for bookmark events and calls fetchHighlights **but** does not broadcast the update.
3. SectionsManager will try to dedupe and looks at `Highlights.rows` but that is not actually the state reflected in the page.
Let me know if this makes sense. My suggestion would be to query the store instead.

**Comment by rlr** [comment]
@piatra That makes sense. But even if fetchHighlights doesn't broadcast, it still calls `SectionManager.updateSection()`: https://github.com/mozilla/activity-stream/blob/master/system-addon/lib/HighlightsFeed.jsm#L119

So that should keep the `SectionManager` data in sync with the main store. But yeah, there might be existing Activity Stream instances not getting updated but they won't get the update from the dedupe either. They just have a stale store but I think that's by design?

**Comment by csadilek** [comment]
@rlr looks good to me! I will do some manual testing as well. @piatra What you describe basically means that existing tabs wouldn't get deduped/updated, right? I think that's fine as in https://github.com/mozilla/activity-stream/pull/3474 we are working on marking the card as "bookmarked" immediately and new tabs from then onwards will get deduped. That matches the requirement if I understand correctly?

**Comment by piatra** [comment]
@csadilek yes that is the requirement. marked as "bookmarked" immediately.

This is kind of extreme but worth mentioning. If you bookmark 10 top stories the first one you bookmarked will surface again in Topstories.

**Comment by csadilek** [comment]
@piatra yes, this makes sense. We only have 9 highlight cards, so after the 10th bookmark, the first one will be back in the feed at its default position (either sorting or scoring in personalization).

**Comment by csadilek** [comment]
@rlr @piatra so this works well in my tests. Happy to approve. You can to decide when to merge it. Before or after beta ;).

**Comment by as-pine-proxy** [comment]
[Treeherder: 2ccedbd9dcba0fdc8e0edde413354f8f55a06d04](https://treeherder.mozilla.org/#/jobs?repo=pine&revision=2ccedbd9dcba0fdc8e0edde413354f8f55a06d04)

**Pull request merged**